datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,829 @@
+///
+/// In order to quickly traverse from heap to map, we use the unsafe raw indexes that `RawTable`
+/// exposes to us to avoid needing to find buckets based on their hash.
+///
+/// Presently this implementation does not traverse quickly from map to heap, instead performing
+/// a B-ary search of the `BTreeSet`. In the future this could be eliminated by implementing a
+/// custom binary heap with pointers from the map into the heap.
+pub struct PriorityMap<ID: KeyType, VAL: ValueType> {
+    limit: usize,
+    desc: bool,
+    rnd: RandomState,
+    id_to_val: RawTable<MapItem<ID, VAL>>,
+    val_to_idx: BTreeSet<HeapItem<VAL>>,
+}
+
+pub struct MapItem<ID: KeyType, VAL: ValueType> {
+    hash: u64,
+    pub id: ID,
+    pub val: VAL,
+}
+
+impl<ID: KeyType, VAL: ValueType> MapItem<ID, VAL> {
+    pub fn new(hash: u64, id: ID, val: VAL) -> Self {
+        Self { hash, id, val }
+    }
+}
+
+struct HeapItem<VAL: ValueType> {
+    val: VAL,
+    buk_idx: usize,
+}
+
+impl<VAL: ValueType> HeapItem<VAL> {
+    pub fn new(val: VAL, buk_idx: usize) -> Self {
+        Self { val, buk_idx }
+    }
+}
+
+impl<VAL: ValueType> Eq for HeapItem<VAL> {}
+
+impl<VAL: ValueType> PartialEq<Self> for HeapItem<VAL> {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl<VAL: ValueType> PartialOrd<Self> for HeapItem<VAL> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl<VAL: ValueType> Ord for HeapItem<VAL> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let res = self.val.compare(other.val);
+        if res != Ordering::Equal {
+            return res;
+        }
+        self.buk_idx.cmp(&other.buk_idx)
+    }
+}
+
+impl<ID: KeyType, VAL: ValueType> PriorityMap<ID, VAL>
+where
+    VAL: PartialEq<VAL>,
+{
+    pub fn new(limit: usize, capacity: usize, desc: bool) -> Self {
+        Self {
+            limit,
+            desc,
+            rnd: Default::default(),
+            id_to_val: RawTable::with_capacity(capacity),
+            val_to_idx: Default::default(),
+        }
+    }
+
+    pub fn insert(&mut self, new_id: ID, new_val: VAL) -> Result<()> {
+        let is_full = self.is_full();
+        let desc = self.desc;
+        assert!(self.len() <= self.limit, "Overflow");
+        let (id_to_val, val_to_idx) = (&mut self.id_to_val, &mut self.val_to_idx);
+
+        // if we're full, and the new val is worse than all our values, just bail
+        if is_full {
+            let worst_entry = if desc {
+                val_to_idx.first()
+            } else {
+                val_to_idx.last()
+            }
+            .expect("Missing value!");
+            if (!desc && new_val > worst_entry.val) || (desc && new_val < worst_entry.val)
+            {
+                return Ok(());
+            }
+        }
+
+        // handle new groups we haven't seen yet
+        let new_hash = self.rnd.hash_one(&new_id);
```

Comment:
I didn't review this implementation carefully yet but the basic idea makes sense to me. I can review it more carefully if @Dandandan  and @thinkharderdev  haven't done so. 

I still believe (without evidence yet) that I can use the same code from https://github.com/apache/arrow-datafusion/pull/7250 and get very similar performance, but I can do that as a follow on PR in parallel with this one